File: app/src/main/java/com/pimenta/bestv/presenter/SearchPresenter.java
/*
 * Copyright (C) 2018 Marcus Pimenta
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package com.pimenta.bestv.presenter;

import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.DisplayMetrics;
import android.util.Log;
import android.util.Pair;
import android.widget.ImageView;

import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.transition.Transition;
import com.pimenta.bestv.BuildConfig;
import com.pimenta.bestv.manager.ImageManager;
import com.pimenta.bestv.repository.MediaRepository;
import com.pimenta.bestv.repository.entity.Movie;
import com.pimenta.bestv.repository.entity.TvShow;
import com.pimenta.bestv.repository.entity.Work;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by marcus on 12-03-2018.
 */
public class SearchPresenter extends BasePresenter<SearchContract> {

    private static final String TAG = SearchPresenter.class.getSimpleName();

    private DisplayMetrics mDisplayMetrics;
    private MediaRepository mMediaRepository;
    private ImageManager mImageManager;

    private String mQuery;
    private int mResultMoviePage = 0;
    private int mResultTvShowPage = 0;
    private Disposable mDisposable;

    @Inject
    public SearchPresenter(DisplayMetrics displayMetrics, MediaRepository mediaRepository, ImageManager imageManager) {
        super();
        mDisplayMetrics = displayMetrics;
        mMediaRepository = mediaRepository;
        mImageManager = imageManager;
    }

    @Override
    public void unRegister() {
        disposeSearchMovie();
        super.unRegister();
    }

    /**
     * Gets the {@link DisplayMetrics} instance
     *
     * @return {@link DisplayMetrics}
     */
    public DisplayMetrics getDisplayMetrics() {
        return mDisplayMetrics;
    }

    /**
     * Searches the movies by a query
     *
     * @param query Query to search the movies
     */
    public void searchWorksByQuery(String query) {
        disposeSearchMovie();
        try {
            String queryEncode = URLEncoder.encode(query, "UTF-8");
            if (!queryEncode.equals(mQuery)) {
                mResultMoviePage = 0;
                mResultTvShowPage = 0;
            }
            mQuery = queryEncode;
            int resultMoviePage = mResultMoviePage + 1;
            int resultTvShowPage = mResultTvShowPage + 1;
            mDisposable = Single.zip(mMediaRepository.searchMoviesByQuery(mQuery, resultMoviePage),
                    mMediaRepository.searchTvShowsByQuery(mQuery, resultTvShowPage),
                    Pair::new)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(pair -> {
                        if (mContract != null) {
                            List<Movie> movies = null;
                            if (pair.first != null && pair.first.getPage() <= pair.first.getTotalPages()) {
                                mResultMoviePage = pair.first.getPage();
                                movies = pair.first.getWorks();
                            }
                            List<TvShow> tvShows = null;
                            if (pair.second != null && pair.second.getPage() <= pair.second.getTotalPages()) {
                                mResultTvShowPage = pair.second.getPage();
                                tvShows = pair.second.getWorks();
                            }
                            mContract.onResultLoaded(movies, tvShows);
                        }
                    }, throwable -> {
                        Log.e(TAG, "Error while searching movies by query", throwable);
                        if (mContract != null) {
                            mContract.onResultLoaded(null, null);
                        }
                    });
        } catch (UnsupportedEncodingException e) {
            Log.e(TAG, "Error while encoding the query", e);
        }
    }

    /**
     * Load the movies by a query
     */
    public void loadMovies() {
        int resultMoviePage = mResultMoviePage + 1;
        mCompositeDisposable.add(mMediaRepository.searchMoviesByQuery(mQuery, resultMoviePage)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(moviePage -> {
                    if (mContract != null) {
                        if (moviePage != null && moviePage.getPage() <= moviePage.getTotalPages()) {
                            mResultMoviePage = moviePage.getPage();
                            mContract.onMoviesLoaded(moviePage.getWorks());
                        } else {
                            mContract.onMoviesLoaded(null);
                        }
                    }
                }, throwable -> {
                    Log.e(TAG, "Error while loading movies by query", throwable);
                    if (mContract != null) {
                        mContract.onMoviesLoaded(null);
                    }
                }));
    }

    /**
     * Load the tv shows by a query
     */
    public void loadTvShows() {
        int resultTvShowPage = mResultTvShowPage + 1;
        mCompositeDisposable.add(mMediaRepository.searchTvShowsByQuery(mQuery, resultTvShowPage)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(tvShowPage -> {
                    if (mContract != null) {
                        if (tvShowPage != null && tvShowPage.getPage() <= tvShowPage.getTotalPages()) {
                            mResultTvShowPage = tvShowPage.getPage();
                            mContract.onTvShowsLoaded(tvShowPage.getWorks());
                        } else {
                            mContract.onTvShowsLoaded(null);
                        }
                    }
                }, throwable -> {
                    Log.e(TAG, "Error while loading tv shows by query", throwable);
                    if (mContract != null) {
                        mContract.onTvShowsLoaded(null);
                    }
                }));
    }

    /**
     * Loads the {@link Work} porter into {@link ImageView}
     *
     * @param work      {@link Work}
     * @param imageView {@link ImageView}
     */
    public void loadWorkPosterImage(@NonNull Work work, ImageView imageView) {
        mImageManager.loadImageInto(imageView, String.format(BuildConfig.TMDB_LOAD_IMAGE_BASE_URL, work.getPosterPath()));
    }

    /**
     * Loads the {@link android.graphics.drawable.Drawable} from the {@link Work}
     *
     * @param work {@link Work}
     */
    public void loadBackdropImage(@NonNull Work work) {
        mImageManager.loadBitmapImage(String.format(BuildConfig.TMDB_LOAD_IMAGE_BASE_URL, work.getBackdropPath()),
                new SimpleTarget<Bitmap>() {
                    @Override
                    public void onResourceReady(@NonNull final Bitmap resource, @Nullable final Transition<? super Bitmap> transition) {
                        if (mContract != null) {
                            mContract.onBackdropImageLoaded(resource);
                        }
                    }

                    @Override
                    public void onLoadFailed(@Nullable final Drawable errorDrawable) {
                        super.onLoadFailed(errorDrawable);
                        Log.w(TAG, "Error while loading backdrop image");
                        if (mContract != null) {
                            mContract.onBackdropImageLoaded(null);
                        }
                    }
                });
    }

    /**
     * Disposes the search movies.
     */
    private void disposeSearchMovie() {
        if (mDisposable != null && !mDisposable.isDisposed()) {
            mDisposable.dispose();
            mDisposable = null;
        }
    }
}

File: app/src/main/java/com/pimenta/bestv/feature/search/presenter/SearchPresenter.kt
/*
 * Copyright (C) 2018 Marcus Pimenta
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package com.pimenta.bestv.feature.search.presenter

import com.pimenta.bestv.common.mvp.AutoDisposablePresenter
import com.pimenta.bestv.common.presentation.model.WorkViewModel
import com.pimenta.bestv.common.usecase.WorkUseCase
import com.pimenta.bestv.extension.addTo
import com.pimenta.bestv.feature.search.usecase.SearchWorksByQueryUseCase
import com.pimenta.bestv.scheduler.RxScheduler
import io.reactivex.Completable
import io.reactivex.disposables.Disposable
import timber.log.Timber
import java.util.concurrent.TimeUnit
import javax.inject.Inject

/**
 * Created by marcus on 12-03-2018.
 */
class SearchPresenter @Inject constructor(
        private val view: View,
        private val workUseCase: WorkUseCase,
        private val searchWorksByQueryUseCase: SearchWorksByQueryUseCase,
        private val rxScheduler: RxScheduler
) : AutoDisposablePresenter() {

    private var query: String = ""
    private var resultMoviePage = 0
    private var resultTvShowPage = 0
    private var searchWorkDisposable: Disposable? = null
    private var loadBackdropImageDisposable: Disposable? = null

    override fun dispose() {
        disposeSearchWork()
        disposeLoadBackdropImage()
        super.dispose()
    }

    fun searchWorksByQuery(text: String) {
        disposeSearchWork()
        resultMoviePage = 0
        resultTvShowPage = 0
        query = text
        searchWorkDisposable = searchWorksByQueryUseCase(text)
                .subscribeOn(rxScheduler.ioScheduler)
                .observeOn(rxScheduler.mainScheduler)
                .subscribe({ pair ->
                    var movies: List<WorkViewModel>? = null
                    if (pair.first.page <= pair.first.totalPages) {
                        this.resultMoviePage = pair.first.page
                        movies = pair.first.works
                    }

                    var tvShows: List<WorkViewModel>? = null
                    if (pair.second.page <= pair.second.totalPages) {
                        this.resultTvShowPage = pair.second.page
                        tvShows = pair.second.works
                    }
                    view.onResultLoaded(movies, tvShows)
                }, { throwable ->
                    Timber.e(throwable, "Error while searching movies by query")
                    view.onResultLoaded(null, null)
                })
    }

    fun loadMovies() {
        workUseCase.searchMoviesByQuery(query, resultMoviePage + 1)
                .subscribeOn(rxScheduler.ioScheduler)
                .observeOn(rxScheduler.mainScheduler)
                .subscribe({ moviePage ->
                    if (moviePage != null && moviePage.page <= moviePage.totalPages) {
                        this.resultMoviePage = moviePage.page
                        view.onMoviesLoaded(moviePage.works)
                    } else {
                        view.onMoviesLoaded(null)
                    }
                }, { throwable ->
                    Timber.e(throwable, "Error while loading movies by query")
                    view.onMoviesLoaded(null)
                }).addTo(compositeDisposable)
    }

    fun loadTvShows() {
        workUseCase.searchTvShowsByQuery(query, resultTvShowPage + 1)
                .subscribeOn(rxScheduler.ioScheduler)
                .observeOn(rxScheduler.mainScheduler)
                .subscribe({ tvShowPage ->
                    if (tvShowPage != null && tvShowPage.page <= tvShowPage.totalPages) {
                        this.resultTvShowPage = tvShowPage.page
                        view.onTvShowsLoaded(tvShowPage.works)
                    } else {
                        view.onTvShowsLoaded(null)
                    }
                }, { throwable ->
                    Timber.e(throwable, "Error while loading tv shows by query")
                    view.onTvShowsLoaded(null)
                }).addTo(compositeDisposable)
    }

    fun countTimerLoadBackdropImage(workViewModel: WorkViewModel) {
        disposeLoadBackdropImage()
        loadBackdropImageDisposable = Completable
                .timer(BACKGROUND_UPDATE_DELAY, TimeUnit.MILLISECONDS)
                .subscribeOn(rxScheduler.ioScheduler)
                .observeOn(rxScheduler.mainScheduler)
                .subscribe({
                    view.loadBackdropImage(workViewModel)
                }, { throwable ->
                    Timber.e(throwable, "Error while loading backdrop image")
                })
    }

    private fun disposeSearchWork() {
        searchWorkDisposable?.run {
            if (!isDisposed) {
                dispose()
            }
        }
    }

    private fun disposeLoadBackdropImage() {
        loadBackdropImageDisposable?.run {
            if (!isDisposed) {
                dispose()
            }
        }
    }

    companion object {

        private const val BACKGROUND_UPDATE_DELAY = 300L
    }

    interface View {

        fun onResultLoaded(movies: List<WorkViewModel>?, tvShows: List<WorkViewModel>?)

        fun onMoviesLoaded(movies: List<WorkViewModel>?)

        fun onTvShowsLoaded(tvShows: List<WorkViewModel>?)

        fun loadBackdropImage(workViewModel: WorkViewModel)

    }
}